python/mspasspy/io/distributed.py

```diff
@@ -0,0 +1,1322 @@
+        objects and produces a single Seismogram from them.  That process
+        can, and usually does, leave things like seed channel names and
+        orientation attributes (hang and vang) from one of the components
+        as extraneous baggage.   Use this of keys to prevent such attributes
+        from being written to the output documents.  Not if the data being
+        saved lack these keys nothing happens so it is safer, albeit slower,
+        to have the list be as large as necessary to eliminate any potential
+        debris.
+    :type exclude_keys: a :class:`list` of :class:`str`
+    :param collection: The default for this parameter is the python
+        None.  The default should be used for all but data export functions.
+        The normal behavior is for this writer to use the object
+        data type to determine the schema is should use for any type or
+        name enforcement.  This parameter allows an alernate collection to
+        be used with or without some different name and type restrictions.
+        The most common use of anything other than the default is an
+        export to a diffrent format.
+    :param data_tag: a user specified "data_tag" key.  See above and
+        User's manual for guidance on how the use of this option.
+    :type data_tag: :class:`str`
+    """
+    # 1. write to file system, distributed
+    t = data.map(
+        lambda cur: write_files(
+            cur, file_format, storage_mode, overwrite, gfsh=gridfs.GridFS(db)
+        )
+    )
+    # 2. write to database, sequential, use map here for concise
+    if format == "spark":
+        r = map(
+            lambda x: write_to_db(
+                db,
+                x,
+                mode=mode,
+                storage_mode=storage_mode,
+                dir=None,
+                dfile=None,
+                format=file_format,
+                overwrite=overwrite,
+                exclude_keys=exclude_keys,
+                collection=collection,
+                data_tag=data_tag,
+                alg_name=alg_name,
+                alg_id=alg_id,
+            ),
+            t.collect(),  # rdd -> list
+        )
+    else:
+        r = map(
+            lambda x: write_to_db(
+                db,
+                x,
+                mode=mode,
+                storage_mode=storage_mode,
+                dir=None,
+                dfile=None,
+                format=file_format,
+                overwrite=overwrite,
+                exclude_keys=exclude_keys,
+                collection=collection,
+                data_tag=data_tag,
+                alg_name=alg_name,
+                alg_id=alg_id,
+            ),
+            t.compute(),  # bag -> list
+        )
+    # convert list of metadata to dataframe
+    return pd.json_normalize(map(lambda cur: cur.todict(), r))
+
+
+def write_to_db(
+    db,
+    mspass_object,
+    mode="promiscuous",
+    storage_mode="gridfs",
```

Comment:
We want to make files the default here. Also, `write_to_db` is called after `write_files`, which means the input already contains the info on how the data is saved. I guess this is because the in `write_files` the return type is still a mspass object, but still the metadata of each mspass object should have the storage mode and other relevant keys (e.g., dir, dfile) defined. Also, I think we should at least make it optional for `write_files` to return dataframe instead, which will make it the exact reverse of `read_files`. Also, `write_to_db` should be a reverse to `read_to_dataframe`.